nrf52840-hal/src/ieee802154.rs

```diff
@@ -0,0 +1,630 @@
+            TxPower::Pos4dBm => TXPOWER_A::POS4DBM,
+            TxPower::Pos5dBm => TXPOWER_A::POS5DBM,
+            TxPower::Pos6dBm => TXPOWER_A::POS6DBM,
+            TxPower::Pos7dBm => TXPOWER_A::POS7DBM,
+            TxPower::Pos8dBm => TXPOWER_A::POS8DBM,
+        }
+    }
+}
+
+impl<'c> Radio<'c> {
+    /// Initializes the radio for IEEE 802.15.4 operation
+    pub fn init<L, LSTAT>(radio: RADIO, _clocks: &'c Clocks<ExternalOscillator, L, LSTAT>) -> Self {
+        let mut radio = Self {
+            radio,
+            _clocks: &(),
+        };
+
+        // go to a known state
+        radio.disable();
+
+        // clear any event interesting to us
+        radio.radio.events_disabled.reset();
+        radio.radio.events_end.reset();
+
+        radio.radio.mode.write(|w| w.mode().ieee802154_250kbit());
+
+        // NOTE(unsafe) radio is currently disabled
+        unsafe {
+            radio.radio.pcnf0.write(|w| {
+                w.s1incl()
+                    .clear_bit() // S1 not included in RAM
+                    .plen()
+                    ._32bit_zero() // 32-bit zero preamble
+                    .crcinc()
+                    .include() // LENGTH field includes CRC (2 bytes)
+                    .cilen()
+                    .bits(0) // no code indicator
+                    .lflen()
+                    .bits(8) // length = 8 bits (but highest bit is reserved and must be `0`)
+                    .s0len()
+                    .clear_bit() // no S0
+                    .s1len()
+                    .bits(0) // no S1
+            });
+
+            radio.radio.pcnf1.write(|w| {
+                w.maxlen()
+                    .bits(Packet::MAX_LEN + 2 /* CRC */) // payload length
+                    .statlen()
+                    .bits(0) // no static length
+                    .balen()
+                    .bits(0) // no base address
+                    .endian()
+                    .clear_bit() // little endian
+                    .whiteen()
+                    .clear_bit() // no data whitening
+            });
+
+            // CRC configuration required by the IEEE spec: x**16 + x**12 + x**5 + 1
+            radio
+                .radio
+                .crccnf
+                .write(|w| w.len().two().skipaddr().ieee802154());
+            radio.radio.crcpoly.write(|w| w.crcpoly().bits(0x11021));
+            radio.radio.crcinit.write(|w| w.crcinit().bits(0));
+
+            // permanent shortcuts
+            radio.radio.shorts.write(|w| w.txready_start().set_bit());
+        }
+
+        // set default settings
+        radio.set_cca(DEFAULT_CCA);
+        radio.set_channel(DEFAULT_CHANNEL);
+        radio.set_sfd(DEFAULT_SFD);
+        radio.set_txpower(DEFAULT_TXPOWER);
+
+        radio
+    }
+
+    /// Changes the radio channel
+    pub fn set_channel(&mut self, channel: Channel) {
+        self.disable();
+
+        // NOTE(unsafe) radio is currently disabled
```

Comment:
It is OK to change frequency while radio is enabled. The FREQUENCY register is double buffered and it is read on RXEN or TXEN tasks.
There is no need to call `self.disable();` before.